### DISCUSSION TITLE
Labs Feedback: @lit-labs/signals

### DISCUSSION THREAD

**Discussion by justinfagnani** [lit/lit#4779] [Labs Feedback]
The `@lit-labs/signals` package is a set utilities for using the [TC39 Signals Proposal](https://github.com/tc39/proposal-signals) [polyfill](https://github.com/proposal-signals/signal-polyfill) in Lit.

- `SignalWatcher`: a class mixin that makes an element watch all signal accesses during the element's reactive update lifecycle, and triggers an element update when signals change.
- `watch()`: a lit-html directive that accepts a single Signal and renders its value, subscribing to updates and updating the DOM when the signal changes.
- `html`, `svg`, and `withWatch()`: `html` and `svg` template tags that can be used in place of Lit's default template tags which automatically wraps any signals in `watch()`. `withWatch()` applies the auto-watching behavior to any lit-html compatible template tag.

We want to collect your feedback on your experiences with `@lit-labs/signals` as we iterate on it according to our Lit Labs process and move the package towards possible graduation or retirement.

- Status: **Prototyping**
- Major known issues: None yet
- Issues: [[labs/signals]](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%signals%5D)
- Docs: [lit.dev/docs/data/signals/](https://lit.dev/docs/data/signals/)
- npm: [@lit-labs/signals](https://www.npmjs.com/package/@lit-labs/signals)

Please try `@lit-labs/signals` and tell us about your experience. Leave comments with your thoughts; highlight what issues or missing features you're blocked on; give a link to a project that uses it; etc.

And so we can have some numbers to look at, you can add reactions to this comment to indicate your experience with this package:

🚀 I have tried this package and it works great. I need it. Ship it!
👍 I have tried this package and it seems like a good idea.
😕 I have tried this package and would use it, but it has serious bugs or missing features.
👀 I am interested, but haven't tried the package yet.
👎 I'm not interested in this package.

Note: other reactions like 😄, 🎉, and ❤️ don't mean anything as part of this scheme. You probably mean, and please use, 🚀, 👍, or  👀.

**Comment by danielrkling** [comment]
Thanks for the work, lit+signals is a great combo. two things
1. Support for watch outside of SignalWatcher. Saw it as a todo comment but definitely needed if I am just using lit-html
2. Avoiding extra wrapping. I see the tension between having watch accept a signal vs a callback as the html wrapper can identify the signal but not a callback vs event, but in my ideal world both watch and the html wrapper for non event parts would work with plain callbacks

**Reply by justinfagnani** [reply]
1. `watch()` needs a scheduler, which lit-html doesn't have yet. We'll have to figure something out there.
2. Wrapping the `html` tag is the way that libraries can hook all expressions. It's really  just an iteration over the values. Something pretty cheap. Bindings already accept plain functions - they pass them on to the property being bound to. This is to support passing functions to components. We can't really change that for the `html` tag.

    `watch()` could conceivably accept a function in addition to a computed, and maybe that sugar is worth it, but there are some perf hazards if people pass function literals in as the function identity will change every render. It's generally better to create the computed outside of `render()` in which case wrapping in `computed(...)` isn't a huge cost.

**Comment by LLT21** [comment]
Would it be possible to make an example with importing from https://cdn.jsdelivr.net/ ?

**Reply by justinfagnani** [reply]
It's an npm package, so you can import it like any npm package form jsdelivr. `https://cdn.jsdelivr.net/npm/@lit-labs/signals/+esm` seems to work.

**Reply by yerlaser** [reply]
What about including this without building?
Lit itself provides bundled js files, can this be done with signals as well?

**Comment by tobi-or-not-tobi** [comment]
@justinfagnani Thank you for this excellent work. I've been testing both this package and `signal-utils`, and everything is working smoothly so far. I'm particularly looking forward to the upcoming features mentioned in the docs, like the repeat directive.

I have a couple of questions about the future direction:
1. Is there a plan to merge the functionality of this package's `html` template tag into the core `html` template tag, or will it remain as a separate feature set?
2. Similarly, will the LitElement (going forward) include `SignalWatcher` capabilities by default? Many applications need these features in most components, and while mixins are a workable solution, adding another mixin layer adds complexity.

Thanks again for your work on this!

**Reply by justinfagnani** [reply]
The reason for being in labs is to try things out before committing to adding them to core. What happens with core depends on the outcome of the labs incubation and maybe more importantly on the TC39 proposal.

I don't see this being integrated into core until Signals ships in some browsers, since it currently relies on a polyfill which can suffer from package duplication issues.

**Reply by tobi-or-not-tobi** [reply]
I appreciate this will not land soon, my questions are more regarding the future: whenever the signals package becomes stable, would it than be part of the core feature set so that we don't need to use an alternative `html` template tag and add the mixin everywhere.

**Comment by justinfagnani** [comment]
Something I noticed today is that it's not necessarily obvious how to compose `watch()` with other directives.

This is how you have to compose `watch()` and `classMap()` for instance:

```ts
class="${watch(computed(() => classMap({foo: count.get() > 0})))}"
```

We might want to make watch wrap any function in a computed to make that simpler:

```ts
class="${watch(() => classMap({foo: count.get() > 0}))}"
```

Though we'd need a way to pass functions. `watch(() => () => ...)`? Or `watch(fn(() => ...)`?

**Comment by Odas0R** [comment]
Very brief comment on the API.

Wouldn't this be more accessible?

```ts
const counter = signal(0);
// current
counter.set(counter.get() + 1);
// more accessible?
counter.set(val => val + 1);
```

Thanks for this implementation, will play with it, and maybe try to contribute!

**Reply by tbroyer** [reply]
This is the Signals API, not the @lit-labs/signals API, do you may want to direct this feedback to https://github.com/tc39/proposal-signals

Fwiw, this is probably so you can set a function as a value in the signal (without having to force use of a callback: `sig.set(_ => f)`)

**Reply by Odas0R** [reply]
> Fwiw, this is probably so you can set a function as a value in the signal (without having to force use of a callback: sig.set(_ => f))

Makes sense!

Thanks :)

**Comment by Odas0R** [comment]
I'm trying to create a reactive binding system for slots in my custom elements. Here's what I'm attempting:

```js
export class Component extends SignalWatcher(LitElement) {
  #bindings = new Map();
  counter = signal(0);

  setupBindings() {
    // Find all slots with data-bind attribute
    const slots = this.querySelectorAll("slot[data-bind]");
    slots.forEach((slot) => {
      const bindingName = slot.dataset.bind;
      if (!bindingName) return;
      // Create signal if it doesn't exist
      if (!this[bindingName]) {
        this[bindingName] = signal("");
      }
      this.#bindings.set(bindingName, slot);
    });
  }

  render() {
    // Attempt to update slots when signals change
    for (const [bindingName, slot] of this.#bindings) {
      render(html`${watch(this[bindingName])}`, slot);
    }
    return this.children;
  }
}
```

In html

```html
<xt-enhance>
  <slot data-bind="counter" />
</xt-enhance>
```

When I update the signal counter (e.g., `counter.set(counter.get() + 1)`), the slot content isn't updating...

But!

If we do 

```js
render() {
  return html`${this.children} ${watch(this.counter)}`;
}
```

It does re-compute that specific DOM element.

Why doesn't the above approach work?

Here's the [playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy5qcyIsImNvbnRlbnQiOiJpbXBvcnQgeyBTaWduYWxXYXRjaGVyLCBzaWduYWwsIHdhdGNoIH0gZnJvbSBcIkBsaXQtbGFicy9zaWduYWxzXCI7XG5pbXBvcnQgeyBMaXRFbGVtZW50LCBodG1sLCByZW5kZXIgfSBmcm9tIFwibGl0XCI7XG5cbi8vIEJhc2UgQ29tcG9uZW50IGNsYXNzIHdpdGggc2lnbmFsIGJpbmRpbmcgZnVuY3Rpb25hbGl0eVxuY2xhc3MgQ29tcG9uZW50IGV4dGVuZHMgU2lnbmFsV2F0Y2hlcihMaXRFbGVtZW50KSB7XG4gICNiaW5kaW5ncyA9IG5ldyBNYXAoKTtcblxuICBjcmVhdGVSZW5kZXJSb290KCkge1xuICAgIHJldHVybiB0aGlzO1xuICB9XG5cbiAgc2V0dXBCaW5kaW5ncygpIHtcbiAgICBjb25zdCBzbG90cyA9IHRoaXMucXVlcnlTZWxlY3RvckFsbChcInNsb3RbZGF0YS1iaW5kXVwiKTtcbiAgICBzbG90cy5mb3JFYWNoKChzbG90KSA9PiB7XG4gICAgICBjb25zdCBiaW5kaW5nTmFtZSA9IHNsb3QuZGF0YXNldC5iaW5kO1xuICAgICAgaWYgKCFiaW5kaW5nTmFtZSkgcmV0dXJuO1xuICAgICAgaWYgKCF0aGlzW2JpbmRpbmdOYW1lXSkge1xuICAgICAgICB0aGlzW2JpbmRpbmdOYW1lXSA9IHNpZ25hbChcIlwiKTtcbiAgICAgIH1cbiAgICAgIHRoaXMuI2JpbmRpbmdzLnNldChiaW5kaW5nTmFtZSwgc2xvdCk7XG4gICAgfSk7XG4gIH1cblxuICByZW5kZXIoKSB7XG4gICAgLy8gRmlyc3QgYXR0ZW1wdDogdHJ5aW5nIHRvIHVwZGF0ZSBzbG90cyBkaXJlY3RseVxuICAgIGZvciAoY29uc3QgW2JpbmRpbmdOYW1lLCBzbG90XSBvZiB0aGlzLiNiaW5kaW5ncykge1xuICAgICAgcmVuZGVyKGh0bWxgJHt3YXRjaCh0aGlzW2JpbmRpbmdOYW1lXSl9YCwgc2xvdCk7XG4gICAgfVxuXG4gICAgLy8gU2Vjb25kIGF0dGVtcHQ6IHRoaXMgd29ya3MgYnV0IGRvZXNuJ3QgdXNlIHNsb3RzXG4gICAgIHJldHVybiBodG1sYFxuICAgICAgICR7dGhpcy5jaGlsZHJlbn0gXG4gICAgICAgSXQgcmUtY29tcHV0ZXM6ICR7d2F0Y2godGhpcy5jb3VudGVyKX1cbiAgICBgO1xuICB9XG5cbiAgY29ubmVjdGVkQ2FsbGJhY2soKSB7XG4gICAgc3VwZXIuY29ubmVjdGVkQ2FsbGJhY2soKTtcbiAgICB0aGlzLnNldHVwQmluZGluZ3MoKTtcbiAgfVxufVxuXG4vLyBFeGFtcGxlIGNvbXBvbmVudCB1c2luZyB0aGUgYmFzZSBjbGFzc1xuY2xhc3MgWHRFbmhhbmNlIGV4dGVuZHMgQ29tcG9uZW50IHtcbiAgY291bnRlciA9IHNpZ25hbCgwKTtcbiAgaW5jKCkge1xuICAgIHRoaXMuY291bnRlci5zZXQodGhpcy5jb3VudGVyLmdldCgpICsgMSk7XG4gICAgY29uc29sZS5sb2coXCJDb3VudGVyIGlzIG5vdzpcIiwgdGhpcy5jb3VudGVyLmdldCgpKTtcbiAgfVxufVxuXG5jdXN0b21FbGVtZW50cy5kZWZpbmUoXCJ4dC1lbmhhbmNlXCIsIFh0RW5oYW5jZSk7XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbjx4dC1lbmhhbmNlPlxuICAgICAgPGJ1dHRvbiB0eXBlPVwiYnV0dG9uXCIgb25jbGljaz1cInRoaXMuY2xvc2VzdCgneHQtZW5oYW5jZScpLmluYygpXCI-XG4gICAgICAgIEluY3JlYXNlXG4gICAgICA8L2J1dHRvbj5cbiAgICAgIEl0IERvZXNuJ3QgcmVjb21wdXRlOiA8c2xvdCBkYXRhLWJpbmQ9XCJjb3VudGVyXCI-PC9zbG90PlxuICA8L3h0LWVuaGFuY2U-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMy4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4yLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeNC4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjMuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d)

**Edit**

Using https://www.npmjs.com/package/@lit-labs/preact-signals it works, so maybe I found a bug?

**Comment by sorvell** [comment]
### Fix

`render(html`${watch(this[bindingName])}`, slot, this.renderOptions);`

The watch directive requires its part's host to be a `SignalWatcher`.

### Warning

If you really want to create a bespoke binding system like this, I'd **strongly** recommend against using the `slot` element.

Using `slot` like this is an **anti-pattern**. It's not being used for its intended purpose inside Shadow DOM and, for example, this was very confusing to me in trying to evaluate your code. It also won't function as it would natively and any changes to the element's children will mess things up.

**Reply by Odas0R** [reply]
Yes, I was tinkering with it and now API has changed, no `slot` no more :)

Thanks for the explanation and advice!

**Reply by Odas0R** [reply]
The idea here is to use custom-elements paired with signals and create a way to add LoB to specific pieces of markup, signals come handy in making the API way simpler

**Comment by tjcrowder** [comment]
Great stuff! 👏🎉

A couple of questions:

1. Is the idea that `html` (and `svg`) in `"lit"` will eventually be replaced with the ones from `@lit-labs/signals`, when all of this is folded into Lit main? Or would they remain distinct even at that point? I ask because the examples in [the blog post](https://lit.dev/blog/2024-10-08-signals/) use the main `html`, not the one from `@lit-labs/signals`, and explicitly use `watch` when that behavior is desired.
2. Suppose a signal is both "watched" in the template and also used directly by the component's code during rendering. Does that avoid double-updates? (Since there's no point in the watch; there will be a re-render when the signal changes.) For a concrete example, consider this in `render` using the base `html` tag function:
   ```js
   const value = someSignal.get();
   return html`<div>A: ${value}</div><div>${watch(someSignal)}</div>;`
   ```

Thanks!

**Reply by justinfagnani** [reply]
> 1. Is the idea that `html` (and `svg`) in `"lit"` will eventually be replaced with the ones from `@lit-labs/signals`, when all of this is folded into Lit main?

It's not definite, but I imagine that graduation for signals might occur in two phases - 1) graduation of the separate signals package (to `@lit/signals`), and 2) folding signals support into Lit core.

When adding signals support into core we can do things a bit differently. We won't need to iterate through values to detect signals, and we won't need the `watch()` directive, we can just make signals a supported data type of all bindings.

> 2. Suppose a signal is both "watched" in the template and also used directly by the component's code during rendering. Does that avoid double-updates? (Since there's no point in the watch; there will be a re-render when the signal changes.) For a concrete example, consider this in `render` using the base `html` tag function:
>    ```js
>    const value = someSignal.get();
>    return html`<div>A: ${value}</div><div>${watch(someSignal)}</div>;`
>    ```

We take care in this package (a change from `@lit-labs/pract-signals`) to batch individual watch directive updates with whole component updates. If there's a whole component update, we just skip the individual updates and update everything. So in this case, the `watch()` is strictly unnecessary, but it won't cause a second render.

**Comment by sorvell** [comment]
We're considering a few behavior changes. These ideas were sparked when considering fixed for https://github.com/lit/lit/issues/4860.

Currently watched signals *do* trigger the Lit reactive update cycle but *do not* provoke Lit's normal rendering and instead perform fine-grained updates. Note that changes to any signals accessed in the lifecycle will also provoke updates (e.g. in `willUpdate` or `updated`). The justification for watched signals explicitly provoking the update cycle, even when they are not otherwise accessed, is to allow code to more sloppily depend on the state of the DOM that may have changed due to a watched signal's fine grained update.

This behavior is complex, error prone, and not obviously especially value. Instead, we're considering simplifying the design so that:
1. watched signals do not trigger reactive updates
2. allow explicitly observing signal changes via an `effect` function that's coordinated with the element's update cycle (e.g. will go after any pending updates).

Open questions
1. should `effect` also run for any reactive update?
2. should `effect` run before the `updateComplete` promise resolves?

**Reply by sorvell** [reply]
In the eng meeting, open questions resolved as:

> 1. should effect also run for any reactive update?

no

> 2. should effect run before the updateComplete promise resolves?

yes, with an option to go before rendering or after, and always before the promise resolves.

We'll also explore a signal based property.

**Comment by mblinsitu** [comment]
Like others, I've been wanting to have `watch` work outside of a `SignalWatcher` component. 

I modified `WatchDirective.__watch` by adding a call to `setTimeout(() => this.commit(), 0)` instead of calling the host's `_updateWatchDirective` when there is no host. I know this is not "the right way" to do it (a post above mentions that there needs to be a scheduler), but so far it has worked for me. Is there something fundamentally wrong with it, or is it just a suboptimal way of doing it?

```js
  private __watch() {
    if (this.__watcher !== undefined) {
      return;
    }
    this.__computed = new Signal.Computed(() => {
      return this.__signal?.get();
    });
    const watcher = (this.__watcher = new Signal.subtle.Watcher(() => {
      // TODO: If we're not running inside a SignalWatcher, we can commit to
      // the DOM independently.
      // this.__host?._updateWatchDirective(this as WatchDirective<unknown>);
      if (this.__host)
        this.__host._updateWatchDirective(this as WatchDirective<unknown>);
      else
        setTimeout(() => this.commit(), 0) // <== new line
      watcher.watch();
    }));
    watcher.watch(this.__computed);
  }
```

**Comment by yaghoubif-ymmd** [comment]
Hi there,

This looks very promising. I have a question , how can I use signals to pass data between components/pages/.... when there is no child/parent relation between ?
Can you make a small example ?
